### PR TITLE
validator

### PR DESCRIPTION
**Pull request by andreineculau**
subset of @lazedo 's #42

### CODE REVIEW

File: src/jesse_state.erl [hunk 1]
```diff
@@ -49,12 +52,16 @@
 %% Includes
 -include("jesse_schema_validator.hrl").
 
+-type extra_validator() :: fun((jesse:json_term(), state()) -> state()) | undefined.
```
Comment (gadgetci) [inline src/jesse_state.erl]: According to **Elvis**:
> Line 55 is too long: -type extra_validator() :: fun((jesse:json_term(), state()) -> state()) | undefined..

File: src/jesse_state.erl [hunk 2]
```diff
@@ -49,12 +52,16 @@
...
+-type setter_fun() :: fun((jesse:json_path(), jesse:json_term(), jesse:json_term()) -> jesse:json_term()) | undefined.
```
Comment (gadgetci) [inline src/jesse_state.erl]: According to **Elvis**:
> Line 56 is too long: -type setter_fun() :: fun((jesse:json_path(), jesse:json_term(), jesse:json_term()) -> jesse:json_term()) | undefined..

File: src/jesse_state.erl [hunk 3]
```diff
@@ -385,3 +406,17 @@ load_schema(#state{schema_loader_fun = LoaderFun}, SchemaURI) ->
...
+
+%% @doc Getter for `current_value'.
+-spec get_current_value(State :: state()) -> jesse:json_term().
+get_current_value(#state{current_value = Value}) -> Value.
+
+-spec set_value(State :: state(), jesse:path(), jesse:json_term()) -> state().
+set_value(#state{setter_fun=undefined}=State, _Path, _Value) -> State;
+set_value(#state{current_value=undefined}=State, _Path, _Value) -> State;
+set_value(#state{setter_fun=Setter
+                ,current_value=Value
```
Comment (gadgetci) [inline src/jesse_state.erl]: According to **Elvis**:
> Missing space right "," on line 418

Comment (gadgetci) [inline src/jesse_state.erl]: According to **Elvis**:
> Missing space right "," on line 418

File: src/jesse_state.erl (hunk 2 above)
Comment (gadgetci) [inline src/jesse_state.erl]: According to **Elvis**:
> Line 56 is too long: -type setter_fun() :: fun((jesse:json_path(), jesse:json_term(), jesse:json_term()) -> jesse:json_term()) | undefined..

File: src/jesse_state.erl (hunk 1 above)
Comment (gadgetci) [inline src/jesse_state.erl]: According to **Elvis**:
> Line 55 is too long: -type extra_validator() :: fun((jesse:json_term(), state()) -> state()) | undefined..

File: src/jesse_state.erl [hunk 4]
```diff
@@ -26,6 +26,7 @@
 %% API
 -export([ add_to_path/2
         , get_allowed_errors/1
+        , get_extra_validator/1
```
Comment (andreineculau) [inline src/jesse_state.erl]: extra/external

File: src/jesse_validator_draft3.erl [hunk 5]
```diff
@@ -209,11 +210,11 @@ check_value(Value, [{?DISALLOW, Disallow} | Attrs], State) ->
 check_value(Value, [{?EXTENDS, Extends} | Attrs], State) ->
   NewState = check_extends(Value, Extends, State),
   check_value(Value, Attrs, NewState);
-check_value(_Value, [], State) ->
-  State;
 check_value(Value, [{?REF, RefSchemaURI} | Attrs], State) ->
   NewState = validate_ref(Value, RefSchemaURI, State),
   check_value(Value, Attrs, NewState);
+check_value(Value, [], State) ->
+  check_external_validation(Value, State);
```
Comment (andreineculau) [inline src/jesse_validator_draft3.erl]: maybe_check_external_validation

File: src/jesse_validator_draft3.erl [hunk 6]
```diff
@@ -48,7 +48,8 @@
                     | ?not_in_range
                     | ?wrong_length
                     | ?wrong_size
-                    | ?wrong_type.
+                    | ?wrong_type
+                    | ?external_error.
```
Comment (andreineculau) [inline src/jesse_validator_draft3.erl]: external

**Pull request merged**